DXMainClient/Domain/Multiplayer/CnCNet/CnCNetTunnel.cs

```diff
@@ -85,6 +85,59 @@ public static CnCNetTunnel Parse(string str)
         public double Distance { get; private set; }
         public int PingInMs { get; set; } = -1;
 
+        /// <summary>
+        /// Rating Factor:
+        ///   The latency in milleseconds that ping times are rounded off to.
+        /// </summary>
+        private const int LATENCY_PRECISION =  70;
+
+        /// <summary>
+        /// Rating Factor:
+        ///   The importance of latency compared to client count.
+        /// </summary>
+        private const int LATENCY_WEIGHT = 100000;
+
+        /// <summary>
+        /// Rating Factor:
+        ///   The value of client count compared to latency
+        /// </summary>
+        private const int CLIENTS_WEIGHT = 1;
+
+        /// <summary>
+        /// Rating Factor:
+        ///   Base Rating for untrusted servers and unpingable servers.
+        /// </summary>
+        private const int UNTRUSTED_RATING = int.MaxValue - 100000;
+
+        /// <summary>
+        /// Rating Factor:
+        ///   The amount of free client slots needed for the tunnel to be eligible for autoselection
+        /// </summary>
+        private const int CLIENTS_HEADROOM = 24;
```

Comment:
Live example: Berryville tunnel has 50 slots, so the server will be running at 26 players max. in case of autochoosing it?
I think the best would be to use tangent function here, so when the tunnel clients amount is very close to max - the rating will skyrocket thus making it never chosen automatically.

Reply:
The CLIENTS_HEADROOM is not for performance reasons. It's meant to make sure that when we go to use a tunnel server it has enough free slots to fill our request. The problem occurs because there is a time difference between when we select the tunnel server (game lobby creation) to when we request the user slots (game start).

Reply:
I understood that already, yet the issue I mentioned is still here. Servers with low max. player amounts would never be chosen or chosen rarely.

IMO the issue of time difference should be dealt with some other sort of measure, f.ex. a task in game lobby checking for availability of the current tunnel on tunnel refresh. Ideally the players in lobby would be counted as players on servers even before starting but that would require server-side changes.